### PR TITLE
Peer details: replace Direction with Connection Type

### LINKED ISSUE #159: Replace "Direction" with "Connection Type" in Peer Tab
To give more detailed information about the connection type, the legacy type (direction) should be replaced by connection type

Current master:

Screenshot from 2020-12-22 12-37-12

### PR DESCRIPTION
**Pull request by jonatack**
Screenshot from 2021-01-09 11-23-17

Closes #159.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Nice. Concept ACK

File: src/qt/guiutil.cpp
Comment (MarcoFalke) [inline src/qt/guiutil.cpp:763]: nit: Would be nice if this default fallback was replaced with an assert(false) and the above ifs with an enum-switch-case

Comment (jonatack) [reply src/qt/guiutil.cpp:763]: Agree, but ISTM that only works for integer types.

Comment (MarcoFalke) [reply src/qt/guiutil.cpp:763]: `ConnectionType` is an enum class (implementation defined integer)

Comment (jonatack) [reply src/qt/guiutil.cpp]: Right, I wanted to do that but the `ConnectionType` enum wasn't exposed in CNodeStats. Ok, done.

**Comment by hebasto** [comment]
Concept ACK.

File: src/qt/guiutil.h
Comment (hebasto) [inline src/qt/guiutil.h:228]: 571b0469fa952be98874d1afd772a7fa708dd83f

This comment is not Doxygen-compatible.

File: src/qt/guiutil.cpp
Comment (hebasto) [inline src/qt/guiutil.cpp:764]: 571b0469fa952be98874d1afd772a7fa708dd83f

The returned values should be translated.

UX note: I have a personal preference do not capitalize the first letter of returned values. Feel free to ignore it.

File: src/qt/guiutil.cpp
Comment (hebasto) [inline src/qt/guiutil.cpp:755]: 571b0469fa952be98874d1afd772a7fa708dd83f

It seems not always true :)

IIUC, you recent changes in `bitcoin-cli` distinguish "block relay" for inbound connections as well.

File: src/qt/forms/debugwindow.ui
Comment (hebasto) [inline src/qt/forms/debugwindow.ui:1083]: bf8135db2a658aed3129ec82e237f8030a8ec7dd

I've noted your [comment](https://github.com/bitcoin-core/gui/pull/162#discussion_r548865476), but here the tooltip provides no additional info. I don't think that adding such a tooltip for the sake itself is a good approach for the GUI:
- https://uxplanet.org/tooltips-in-ui-design-f63e117aa3d1
- https://www.nngroup.com/articles/tooltip-guidelines/

**Review by hebasto — CHANGES_REQUESTED** [review]
Concept ACK.

I have reviewed the code (bf8135db2a658aed3129ec82e237f8030a8ec7dd), and tested it.

File: src/qt/guiutil.cpp
Comment (jonatack) [reply src/qt/guiutil.cpp:764]: I also prefer all lowercase (e.g. rpc/cli) but the GUI UI uses capitalization (I hesitated to use titlecase here).

File: src/qt/forms/debugwindow.ui
Comment (jonatack) [reply src/qt/forms/debugwindow.ui:1083]: True. Perhaps it should link to more information about the connection types (like we have in the codebase), maybe in a markdown file in `/doc`.

File: src/qt/guiutil.cpp
Comment (jonatack) [reply src/qt/guiutil.cpp:755]: > IIUC, you recent changes in `bitcoin-cli` distinguish "block relay" for inbound connections as well.

Yes, I'm anticipating `inbound full relay` and `inbound block relay` connection types (https://github.com/bitcoin/bitcoin/pull/20726 and https://github.com/bitcoin/bitcoin/pull/20729) but I guess it's better to revert to just inbound for now and update it later.

File: src/qt/forms/debugwindow.ui
Comment (jonatack) [reply src/qt/forms/debugwindow.ui:1083]: Maybe I should add more useful info, like in the RPC help (getpeerinfo)

File: src/qt/guiutil.cpp
Comment (jonatack) [reply src/qt/guiutil.cpp:755]: (in the latest -netinfo update in https://github.com/bitcoin/bitcoin/pull/20764, the issue is avoided by using the connection type only to parse full/block/manual/feeler/addr, but not for the direction)

File: src/qt/guiutil.cpp
Comment (hebasto) [reply src/qt/guiutil.cpp:764]: I perceive it like "Connection type: Outbound full relay". That's why I prefer the lowercased variant.

If the capitalized variant will be required in some other places in the GUI, it is easy to provide it with a generic function.

I'm not a design expert, then I'm fine with any variant.

File: src/qt/guiutil.cpp
Comment (jonatack) [reply src/qt/guiutil.cpp:755]: In the latest -netinfo update in https://github.com/bitcoin/bitcoin/pull/20764, the issue is avoided by using the connection type only to parse full/block/manual/feeler/addr, but not for the direction. Maybe best to do the same here, hm.

File: src/qt/forms/debugwindow.ui
Comment (hebasto) [reply src/qt/forms/debugwindow.ui:1083]: Anyway, please keep it useful and _succinct_, or else do not add it at all :)

File: src/qt/guiutil.h
Comment (jonatack) [reply src/qt/guiutil.h:228]: thanks! fixed

File: src/qt/guiutil.cpp
Comment (jonatack) [reply src/qt/guiutil.cpp:764]: Thanks--translated and changed to titlecase for coherence with the general GUI UI.

File: src/qt/guiutil.cpp
Comment (jonatack) [reply src/qt/guiutil.cpp:755]: Now "Inbound"

File: src/qt/forms/debugwindow.ui
Comment (jonatack) [reply src/qt/forms/debugwindow.ui:1083]: Added info similar to the getpeerinfo help, see updated screenshot in the PR description.

**Comment by jonatack** [comment]
Thanks for the excellent feedback @hebasto! Updated.

**Comment by hebasto** [comment]
I have reviewed the code (8e7f82a6a84c7b2553f88767dd356f480d1ae4b7), and tested it.

The following patch:
```diff
--- a/src/qt/forms/debugwindow.ui
+++ b/src/qt/forms/debugwindow.ui
@@ -1079,7 +1079,7 @@
                <item row="1" column="0">
                 <widget class="QLabel" name="peerConnectionTypeLabel">
                  <property name="toolTip">
-                  <string>The type of peer connection: Outbound Full Relay (default), Outbound Block Relay (does not relay transactions or addresses), Inbound (initiated by peer), Outbound Manual (added using RPC addnode or -addnode/-connect config options), Outbound Feeler (short-lived, for testing addresses), or Outbound Address Fetch (short-lived, for soliciting addresses).</string>
+                  <string>The type of peer connection: Outbound Full Relay (default), Outbound Block Relay (does not relay transactions or addresses), Inbound (initiated by peer), Outbound Manual (added using RPC %1 or %2/%3 config options), Outbound Feeler (short-lived, for testing addresses), or Outbound Address Fetch (short-lived, for soliciting addresses).</string>
                  </property>
                  <property name="text">
                   <string>Connection Type</string>
--- a/src/qt/rpcconsole.cpp
+++ b/src/qt/rpcconsole.cpp
@@ -496,6 +496,8 @@ RPCConsole::RPCConsole(interfaces::Node& node, const PlatformStyle *_platformSty
     clear();
 
     GUIUtil::handleCloseWindowShortcut(this);
+
+    ui->peerConnectionTypeLabel->setToolTip(ui->peerConnectionTypeLabel->toolTip().arg("addnode").arg(QString(nonbreaking_hyphen) + "addnode").arg(QString(nonbreaking_hyphen) + "connect"));
 }
 
 RPCConsole::~RPCConsole()
```

- removes RPC and option names from the translatable string
- prevents effects like this
DeepinScreenshot_select-area_20201225232751

**Comment by jonatack** [comment]
Thanks @hebasto, tested and added your suggestion, updated the screenshot in the PR description.

**Review by hebasto — APPROVED** [review]
ACK 6cd242c50f4a8d3f3c509937d80962fe58096ead, tested on Linux Mint 20 (x86_64) + Qt 5.12.8.

Looks great:
Screenshot from 2020-12-25 23-52-27

File: src/net.h
Comment (promag) [inline src/net.h:728]: a60f67b2fb0eede205db8cce11ad0e6f42dbc90e

This was suggested here https://github.com/bitcoin/bitcoin/pull/19725#discussion_r473361238 and now makes 
`m_conn_type_string` redundant, don't makes sense to keep both TBH. Since `m_conn_type_string` is only used in `getpeerinfo` then drop it and refactor `ConnectionTypeAsString` to take `conn_type` instead?

Comment (MarcoFalke) [reply src/net.h:728]: Concept ACK, but RPC changes (refactoring) can only be made in the main repo

Comment (jonatack) [reply src/net.h:731]: Done, I think. My initial version of this PR did not touch any non-GUI code. Should I close this PR and re-open in the main repo? (I'm not sure it's worth losing the review above.)

Comment (hebasto) [reply src/net.h:731]: Maybe just open a pr in the main repo with the first and last commits?

Comment (promag) [reply src/net.h:731]: IMO this isn't a RPC change, it's just a member refactor in a net struct, and the refactor is motivated by this GUI change. Otherwise this can go forward in 2 ways I think, either leave this as it was, and then in main repo drop the redundant member, or 1st refactor and make this dependant of that PR.

Comment (jonatack) [reply src/net.h:731]: Yes, this doesn't change the RPC API.

Please let me know which option is best:

1. split it into two pulls, one here, gated by a another one in the other repo
2. move the pull to the other repo
3. revert to the original version that didn't touch main repo code
4. leave this as-is

Comment (jonatack) [reply src/net.h:731]: I'm not sure the refactoring changes make sense without the whole. Opened the same pull in https://github.com/bitcoin/bitcoin/pull/20778 and will let the maintainers decide.

Comment (promag) [reply src/net.h:731]: Just noting that redundancy is not a blocker, but could be avoided.

**Comment by jonatack** [comment]
It was suggested to re-open here.

**Comment by MarcoFalke** [comment]
Concept re-ACK. Thanks for re-opening.

**Comment by jonatack** [comment]
Rebased.

**Review by hebasto — APPROVED** [review]
re-ACK aef969363a7685fc805833a3ec6863c27500d02f, since my [previous](https://github.com/bitcoin-core/gui/pull/163#pullrequestreview-558838006) review only rebased and the _"p2p, rpc, refactor: remove CNodeStats::m_conn_type_string"_ commit added.

**Comment by jonatack** [comment]
Rebased `git range-diff ae8f797 aef9693 7098509`

File: src/rpc/net.cpp
Comment (hebasto) [inline src/rpc/net.cpp:252]: 70985097c00827de94a7922620b63e1ee10dd53d

Mind fixing https://github.com/bitcoin/bitcoin/pull/19725/commits/395acfa83a5436790c1a722a5609ac9d48df235f (https://github.com/bitcoin/bitcoin/pull/19725) by aligning the order of the "connection_type" field both in RPC help and in RPC output?

**Review by hebasto — APPROVED** [review]
re-ACK 70985097c00827de94a7922620b63e1ee10dd53d, only rebased since my [previous](https://github.com/bitcoin-core/gui/pull/163#pullrequestreview-560126996) review.

File: src/rpc/net.cpp
Comment (jonatack) [reply src/rpc/net.cpp:252]: thanks; done

**Comment by jonatack** [comment]
Thanks @hebasto, updated the last commit to move the getpeerinfo connection_type help to the correct place per review feedback in https://github.com/bitcoin-core/gui/pull/163#discussion_r550863551 and updated the commit message.

**Review by hebasto — APPROVED** [review]
re-ACK be8294ff428da7d13043b3f9a27d7ace1c2dadf4

**Comment by laanwj** [comment]
Concept ACK

**Comment by jonatack** [comment]
Rebased after the merge of https://github.com/bitcoin/bitcoin/pull/20786.

**Review by hebasto — APPROVED** [review]
re-ACK 20fa97560b2d27f94f556a570f3556356877b79f, only rebased since my [previous](https://github.com/bitcoin-core/gui/pull/163#pullrequestreview-560723519) review due to the conflict with https://github.com/bitcoin/bitcoin/pull/20786, verified with
```
$ git range-diff master be8294ff428da7d13043b3f9a27d7ace1c2dadf4 20fa97560b2d27f94f556a570f3556356877b79f
```

**Review by jarolrod — COMMENTED** [review]
ACK 20fa97560b2d27f94f556a570f3556356877b79f

Tested on macOS Big Sur 11.1 with qt 5.15.2. Looks great!
Screen Shot 2021-01-08 at 11 01 46 AM

Small NIT: There's a lot of text in the tooltip. It would be nice if all the possible values were put into a list instead of separated by commas. It would look something like [this](https://ptgmedia.pearsoncmg.com/images/chap17_9780132354165/elementLinks/catalogentry.jpg), which to me would be an improvement.

See: [Supported HTML Subset](https://doc.qt.io/qt-5/richtext-html-subset.html),

**Comment by jonatack** [comment]
Better? If yes, bold or no?
Screenshot from 2021-01-09 00-04-04
Screenshot from 2021-01-08 23-57-49

**Comment by jarolrod** [comment]
@jonatack the list looks great! I think not-bold is better. If i may, I would suggest one more thing:

Write the descriptions like this:
`
Outbound Block Relay: does not relay transactions or addresses
`

instead of:
`
Outbound Block Relay (does not relay transactions or addresses)
`

**Comment by jonatack** [comment]
@jarolrod test with inbound first and your colon suggestion (agree, it's more readable)

Screenshot from 2021-01-09 01-10-03

**Comment by jarolrod** [comment]
@jonatack looks awesome 👍

**Comment by jonatack** [comment]
Thanks @jarolrod. Updated in latest push, if you'd like to review/test. @hebasto, can you verify the changed tooltip code?

**Comment by hebasto** [comment]
> @hebasto, can you verify the changed tooltip code?

The used HTML code is [valid](https://doc.qt.io/qt-5/richtext-html-subset.html).

But I'd prefer to keep the tooltip content in the form `*.ui` file. It makes usage of design tools (e.g., Qt Designer) much easier. In `*.cpp` file should be text substitution code only.

**Comment by jonatack** [comment]
@hebasto thanks--done.

**Review by hebasto — APPROVED** [review]
re-ACK 06ba9b300866f33e21512af9d7d2891ee1501bf4, the tooltip content is organized as unordered list.

**Review by jarolrod — COMMENTED** [review]
re-ACK 06ba9b300866f33e21512af9d7d2891ee1501bf4

File: src/qt/forms/debugwindow.ui
Comment (MarcoFalke) [inline src/qt/forms/debugwindow.ui:1082]: translation nit: How are translators supposed to translate `:&lt;ul&gt;&lt;li&gt;`? It might be better to say "Can be one of %s" and then use `Join(translated_strings, ", ")` or simply "Refer to the getpeerinfo RPC help for details on the connection types"?

Comment (jonatack) [reply src/qt/forms/debugwindow.ui:1082]: @hebasto, if we replace the escaped markup with placeholders (as done here for the nonbreaking hyphen), is that still QT Designer-friendly? Can do it in #179.

Comment (hebasto) [reply src/qt/forms/debugwindow.ui:1082]: > translation nit: How are translators supposed to translate `:&lt;ul&gt;&lt;li&gt;`?

There was a discussion in private IRC with @jonatack:
> \<hebasto\>
> 14:24:52 but I expect the translators' burden :)
> 14:25:39 hope that translators are aware of escaping :)

I'm not an expert in translation but I believe that `:&lt;ul&gt;&lt;li&gt;` is not a subject to translate as `%1` or `%2`.

> It might be better to say "Can be one of %s" and then use `Join(translated_strings, ", ")` or simply "Refer to the getpeerinfo RPC help for details on the connection types"?

Either of the suggested variants LGTM.

> if we replace the escaped markup with placeholders (as done here for the nonbreaking hyphen), is that still QT Designer-friendly?

I think it is. But, tbh, I'd prefer more succinct @MarcoFalke's suggestions.

Comment (jonatack) [reply src/qt/forms/debugwindow.ui:1082]: Thanks. Helpfulness and readability for users were the reasons for the iterations above and it seems best to not regress on those improvements with referring GUI users to the command line or dropping the readable list layout. I'll propose replacing the escaped markup with placeholders.

Comment (hebasto) [reply src/qt/forms/debugwindow.ui:1082]: Why not a single placeholder for the whole unordered list? We still have a (shorter) tooltip in the form file, and make translators' life easier :)

Comment (jonatack) [reply src/qt/forms/debugwindow.ui:1082]: SGTM too.

Comment (jonatack) [reply src/qt/forms/debugwindow.ui:1082]: Done in #180

**Pull request merged**
